spec/controllers/users_controller_spec.rb

```diff
@@ -120,34 +120,78 @@
-          expect(assigns(:user)).to eq(user)
-        end
+        context "communication opt-in" do
+          before do
+            Timecop.freeze(Time.new(2002, 10, 31))
+          end
+
+          after do
+            Timecop.return
+          end
 
-        it "re-renders the 'edit' template" do
-          allow_any_instance_of(User).to receive(:save).and_return(false)
-          put :update, params: { id: user.to_param, user: { name: 'invalid value' } }
-          expect(response).to render_template("edit")
+          shared_examples_for 'tracks opt-in time' do |attribute|
+            it "sets #{attribute} to the current time" do
+              put :update, params: { id: user.to_param, user: { attribute => '1' } }
+              expect(user.reload.send "#{attribute}_at").to eq(Time.new(2002, 10, 31))
+            end
+          end
+
+          it_behaves_like 'tracks opt-in time', :opted_in_newsletter
+          it_behaves_like 'tracks opt-in time', :opted_in_announcements
+          it_behaves_like 'tracks opt-in time', :opted_in_marketing_announcements
+          it_behaves_like 'tracks opt-in time', :opted_in_surveys
+          it_behaves_like 'tracks opt-in time', :opted_in_sponsorships
+          it_behaves_like 'tracks opt-in time', :opted_in_applications_open
+
+          context 'opting out' do
+            shared_examples_for 'tracks opt-out' do |attribute|
+              before do
+                user.update_attribute attribute, Time.new(2002, 10, 31)
+              end
+
+              it "sets #{attribute} to nil" do
+                put :update, params: { id: user.to_param, user: { attribute => '0' } }
+                expect(user.reload.send "#{attribute}_at").to eq(nil)
+              end
+            end
+
+            it_behaves_like 'tracks opt-out', :opted_in_newsletter
+            it_behaves_like 'tracks opt-out', :opted_in_announcements
+            it_behaves_like 'tracks opt-out', :opted_in_marketing_announcements
+            it_behaves_like 'tracks opt-out', :opted_in_surveys
+            it_behaves_like 'tracks opt-out', :opted_in_sponsorships
+            it_behaves_like 'tracks opt-out', :opted_in_applications_open
+          end
         end
       end
+    end
 
-      context "another user's profile" do
-        let!(:another_user) { create(:user) }
+    describe "with invalid params" do
+      it "assigns the user as @user" do
+        allow_any_instance_of(User).to receive(:save).and_return(false)
+        put :update, params: { id: user.to_param, user: { name: 'invalid value' } }
+        expect(assigns(:user)).to eq(user)
+      end
 
-        it "does not update the requested user" do
-          expect_any_instance_of(User).not_to receive(:update_attributes)
-          put :update, params: { id: another_user.to_param, user: { name: 'Trung Le' } }
-        end
+      it "re-renders the 'edit' template" do
+        allow_any_instance_of(User).to receive(:save).and_return(false)
```

Comment:
Personal preference, but I'd suggest to not use `allow_any_instance_of` for cases where it's easy to just have _a real world scenario_. So if you just update the params to result in a validation error, things are even tested **better**, since you don't have to know the internals of the controller, such as if it calls `save` or `save!` _(for which this test would fail btw.). So maybe just leave the name blank or something like this and get rid of the `allow_any_instance` bit.